src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -425,14 +440,114 @@ private static double getRangesTotalSize(List<Splitter.WeightedRange> ranges)
             ++perLevel[level];
             ++runningCompactions;
             levelCount = Math.max(levelCount, level + 1);
-            spaceAvailable -= compaction.totSizeInBytes;
+            spaceAvailable -= compaction.totSizeInBytes();
         }
 
+        CompactionLimits limits = new CompactionLimits(runningCompactions, 
+                                                       maxCompactions,
+                                                       maxConcurrentCompactions,
+                                                       perLevel,
+                                                       levelCount,
+                                                       spaceAvailable,
+                                                       rateLimitLog);
         logger.debug("Selecting up to {} new compactions of up to {}, concurrency limit {}{}",
-                     maxCompactions - runningCompactions,
-                     FBUtilities.prettyPrintMemory(spaceAvailable),
-                     maxConcurrentCompactions,
-                     rateLimitLog);
+                     Math.max(0, limits.maxCompactions - limits.runningCompactions),
+                     FBUtilities.prettyPrintMemory(limits.spaceAvailable),
+                     limits.maxConcurrentCompactions,
+                     limits.rateLimitLog);
+        return limits;
+    }
+
+    /**
+     * Selects compactions to run next.
+     * 
+     * @param gcBefore
+     * @return a subset of compaction aggregates to run next
+     */
+    private Collection<CompactionAggregate> getNextCompactionAggregates(int gcBefore)
```

Comment:
I see - I did not feel very comfortable to break up the `for` statement and have to iterate twice in the one case. To me it also makes the adjustment to `limits.levelCount` less clear because they are no longer done together.
I am talking about this code that would be split with the proposed change:
```
{
    // ... and also the levels that a layout-preserving selection would create.
    limits.levelCount = Math.max(limits.levelCount, levelOf(selected) + 1);
    limits.levelCount = Math.max(limits.levelCount, (int) selected.parent());
}
```

Reply:
> Do I understand correctly that the added test is testing `getNextCompactionAggregates` for no compaction? Shoudl it be a test when the result of the method should not be empty?

I will change it

Reply:
> I see - I did not feel very comfortable to break up the `for` statement and have to iterate twice in the one case.

Regarding the extra loop, note that your PR already adds another loop in comparison with the original code, where `levelCount` was updated at the same time when the aggregation was added to the pending list.